app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/upscan/UpscanController.scala

```diff
@@ -83,56 +84,101 @@ class UpscanController @Inject() (
-            case Left(error) =>
-              logger.warn(s"Could not upload file to S3", error)
-              errorHandler.errorResult(None)
-            case Right(_) => Ok
-          }
-      }
-  }
-
-  def callBack(cgtReference: String): Action[JsValue] = Action.async(parse.json) { implicit request =>
-    request.body
-      .validate[UpscanNotifyResponse]
-      .asOpt
-      .fold(Future.successful(BadRequest("Invalid call back response received from upscan notify")))(notifyEvent =>
-        upscanService.storeNotifyEvent(CgtReference(cgtReference), notifyEvent).value.map {
+  def upscan(): Action[AnyContent] =
+    authenticatedActionWithSessionData.async { implicit request =>
+      withSubscribedUser(request) { (_, subscribed) =>
+        upscanService.initiate(subscribed.subscribedDetails.cgtReference).value.map {
           case Left(error) =>
-            logger.warn("upscan notifier call back handler failed", error)
+            logger.warn(s"could not initiate upscan due to $error")
             errorHandler.errorResult(None)
-          case Right(_) =>
-            logger.info("upscan notifier call back succeeded")
-            NoContent
+          case Right(upscanInitiateResponse) =>
+            upscanInitiateResponse match {
+              case FailedToGetUpscanSnapshot                => errorHandler.errorResult(None)
+              case MaximumFileUploadReached                 => Ok(upscanLimitPage())
+              case UpscanInititateResponseStored(reference) => Ok(upscanPage(reference))
+            }
         }
-      )
-  }
+      }
+    }
 
-  def upscan(): Action[AnyContent] = authenticatedActionWithSessionData.async { implicit request =>
-    withSubscribedUser(request) { (_, subscribed) =>
-      upscanService.initiate(subscribed.subscribedDetails.cgtReference).value.map {
-        case Left(error) =>
-          logger.warn(s"Could not initiate upscan due to $error")
-          errorHandler.errorResult(None)
-        case Right(upscanResult) =>
-          upscanResult match {
-            case MaximumFileUploadReached            => Ok(upscanLimitPage())
-            case UpscanResponse(_, upscanDescriptor) => Ok(upscanPage(upscanDescriptor))
+  @SuppressWarnings(Array("org.wartremover.warts.Var", "org.wartremover.warts.Any"))
+  def upload(): Action[MultipartFormData[Files.TemporaryFile]] =
+    authenticatedActionWithSessionData(parse.multipartFormData(maxFileSize)).async { implicit request =>
+      withSubscribedUser(request) { (_, subscribed) =>
+        val multipart: MultipartFormData[Files.TemporaryFile] = request.body
+        multipart.dataParts
+          .get("reference")
+          .flatMap(_.headOption)
+          .fold(Future.successful(BadRequest("missing upscan file descriptor id"))) { reference =>
+            upscanService
+              .getUpscanFileDescriptor(subscribed.subscribedDetails.cgtReference, UpscanInitiateReference(reference))
+              .value
+              .flatMap {
```

Comment:
could this not be written in a for-compresion?

Reply:
also, can any of this be orchestrated by the BE? would it makes sense for the FE to do a single call and have the BE do the heavy lifting?

Reply:
> could this not be written in a for-compresion?

Yes I suppose, I was thrashing it out at the time. I will refactor it.

Reply:
> also, can any of this be orchestrated by the BE? would it makes sense for the FE to do a single call and have the BE do the heavy lifting?

yeah we could do that - I will give it some more thought.

Reply:
refactored as a for-comp